frontend/src/metabase/common/components/EntityPicker/components/EntityPickerModal/EntityPickerModal.tsx

```diff
@@ -119,53 +135,120 @@ export function EntityPickerModal<
 
   const { open } = useModalOpen();
 
-  const tabModels = useMemo(
-    () => passedTabs.map(t => t.model).filter(Boolean),
-    [passedTabs],
-  );
+  const searchModels = useMemo(() => getSearchModels(passedTabs), [passedTabs]);
+  const folderModels = useMemo(() => getFolderModels(passedTabs), [passedTabs]);
 
   const filteredRecents = useMemo(() => {
-    const relevantModelRecents =
-      recentItems?.filter(recentItem =>
-        tabModels.includes(recentItem.model as Model),
-      ) || [];
+    if (!recentItems) {
+      return [];
+    }
+
+    const relevantModelRecents = recentItems.filter(recentItem => {
+      return searchModels.includes(recentItem.model);
+    });
 
     return recentFilter
       ? recentFilter(relevantModelRecents)
       : relevantModelRecents;
-  }, [recentItems, tabModels, recentFilter]);
+  }, [recentItems, searchModels, recentFilter]);
+
+  const tabs: EntityPickerTab<Id, Model, Item>[] = useMemo(() => {
+    const computedTabs: EntityPickerTab<Id, Model, Item>[] = [];
+    const hasRecentsTab =
+      hydratedOptions.hasRecents && filteredRecents.length > 0;
+    const hasSearchTab = !!searchQuery;
+    // This is to prevent different tab being initially open and then flickering back
+    // to recents tab once recents have loaded (due to computeInitialTab)
+    const shouldOptimisticallyAddRecentsTabWhileLoading =
+      defaultToRecentTab && isLoadingRecentItems;
+
+    if (hasRecentsTab || shouldOptimisticallyAddRecentsTabWhileLoading) {
```

Comment:
This `shouldOptimisticallyAddRecentsTabWhileLoading` is new, and is not related to the PR or epic in any way, but it's been annoying me since day 1 and it made it really difficult to reason which tab is actually opened initially.